dhis-2/dhis-api/src/main/java/org/hisp/dhis/programrule/engine/RuleActionImplementer.java

```diff
@@ -43,4 +43,8 @@
     void implement( RuleEffect ruleEffect, ProgramInstance programInstance );
 
     void implement( RuleEffect ruleEffect, ProgramStageInstance programStageInstance );
+
+    void implementEnrollmentAction( RuleEffect ruleEffect, String programInstance );
+
+    void implementEventAction( RuleEffect ruleEffect, String programStageInstance );
```

Comment:
This interface is a bit inconistent, since it uses Objects and String as function parameters for `ProgramInstance` and `ProgramStageInstance`

Reply:
yes i agree. The reason is these implementer classes were previously invoked only by ProgramRuleService ( which works in async way ). But in this feature, i am already receiving ProgramRule effects in TrackerBundle object and all i have to do is to invoke implementer classes to apply those effects. This invocation is happening in SideEffectHandlerService and it uses parallel stream. So in order to avoid those lazy loading problems, i have created two new methods to take UIDs and explicitly fetch those objects.